rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -364,6 +364,20 @@ bool RecorderImpl::is_paused()
 
 void RecorderImpl::topics_discovery()
 {
+  // If using sim time - wait until /clock topic received before even creating subscriptions
+  if (record_options_.use_sim_time) {
```

Comment:
The problem here is the initial batch of potential subscriptions in `record()`, before starting the `topics_discovery` thread. Unfortunately we do need `record` to return immediately so I think I will need to put the clock waiting process in its own `std::async` call from `record`. This block of code is approximately correct, but some subscriptions can be created before it is called.

Reply:
@emersonknapp I had a bit more simple fix in my mind to add a check 
```
if (record_options_.use_sim_time) {
    subscribe_topics(get_requested_or_available_topics());
  }
```
in the `Recorder::record()` method. i.e. not subscribe to the specified topics on record and let spawned discovery thread do this with the delay as you did in the original PR.
It will have a side effect that `record_options_.is_discovery_disabled` and `record_options_.use_sim_time` will became incompatible if they are set to true together. 
However, we can add another check and throw a warning with an exception for invalid parameters settings.

Reply:
So you're saying `RecordOptions::is_discovery_disabled && RecordOptions::use_sim_time` should be an invalid setting. OK, that sounds like a good idea to me. It is reasonable to say something like "we have to discover the /clock topic" (even if this doesn't 100% capture the nuance of the implementation). I will add an error check for that configuration case.